test/buy_and_activate.js

```diff
@@ -46,22 +57,153 @@ contract('Court presale and activate wrapper', ([_, owner, juror1, juror2, juror
+
+        beforeEach('deploy external token, mint and approve', async () => {
+          externalToken = await ERC20.new('External Token', 'ET', 18)
+          await externalToken.mint(juror1, INITIAL_BIG_TOKEN_AMOUNT)
+        })
+
+        it('fails when amount is zero', async () => {
+          await assertRevert(cpa.contributeExternalToken(externalToken.address, juror1, 0, 1, 1, 0), ERROR_ZERO_AMOUNT)
+        })
+
+        it('buys, stakes and activates', async () => {
+          const externalExchangeRate = bn(500000)
+          await uniswapExchange.setExchangeRate(externalToken.address, externalExchangeRate)
+          const collateralAmount = DEFAULTS.minActiveBalance.mul(PPM).div(exchangeRate)
+          const externalAmount = collateralAmount.mul(PPM).div(externalExchangeRate)
+          const initialActiveAmount = (await registry.balanceOf(juror1))[0]
+          const bondedTokensToGet = await presale.contributionToTokens(collateralAmount);
+
+          await externalToken.approve(cpa.address, externalAmount, { from: juror1 })
+          await cpa.contributeExternalToken(externalToken.address, juror1, externalAmount, 1, 1, 0, { from: juror1 })
+
+          const finalActiveAmount = (await registry.balanceOf(juror1))[0]
+          assertBn(finalActiveAmount, initialActiveAmount.add(bondedTokensToGet), `Active balance `)
+        })
+      })
+
+      context('ETH', () => {
+        const ETH = '0x' + '0'.repeat(40)
+
+        it('fails when amount is zero', async () => {
+          await assertRevert(cpa.contributeEth(1, 0, { from: juror1, value: 0 }), ERROR_ZERO_AMOUNT)
+        })
+
+        it('buys, stakes and activates', async () => {
+          const ethExchangeRate = bn(50000000000)
+          await uniswapExchange.setExchangeRate(ETH, ethExchangeRate)
+          const collateralAmount = DEFAULTS.minActiveBalance.mul(PPM).div(exchangeRate)
+          const ethAmount = collateralAmount.mul(PPM).div(ethExchangeRate)
+          const initialActiveAmount = (await registry.balanceOf(juror1))[0]
+          const bondedTokensToGet = await presale.contributionToTokens(collateralAmount);
+
+          await cpa.contributeEth(1, 0, { from: juror1, value: ethAmount })
+
+          const finalActiveAmount = (await registry.balanceOf(juror1))[0]
+          assertBn(finalActiveAmount, initialActiveAmount.add(bondedTokensToGet), `Active balance `)
+        })
+      })
+    })
+  })
+
+  context('Bad tokens', () => {
+    let cpa, badCollateralToken, badExternalToken
+
+    beforeEach('Deploy airdrop contract', async () => {
+      // deploy bad tokens
+      badCollateralToken = await ERC20Bad.new('Collateral Bad Token', 'BCT', 18)
+      await badCollateralToken.mint(juror1, INITIAL_BIG_TOKEN_AMOUNT)
+      badExternalToken = await ERC20Bad.new('External Token', 'ET', 18)
+      await badExternalToken.mint(juror1, INITIAL_BIG_TOKEN_AMOUNT)
+
+      // add it to Presale
+      await presale.setCollateralToken(badCollateralToken.address)
```

Comment:
Hmm, why not set up the test from scratch using `badCollateralToken` in the setup for the presale?

Reply:
Yes, I'll do it, see comment above.